go/common/batches.go

```diff
@@ -10,12 +12,22 @@ type ExtBatch struct {
 	Header          *Header
 	TxHashes        []TxHash // The hashes of the transactions included in the batch.
 	EncryptedTxBlob EncryptedTransactions
+	hash            atomic.Value
 }
 
-// TODO - #718 - Cache hash calculation.
+// Hash returns the keccak256 hash of the batch's header.
+// The hash is computed on the first call and cached thereafter.
+func (r *ExtBatch) Hash() L2RootHash {
+	if hash := r.hash.Load(); hash != nil {
+		return hash.(L2RootHash)
+	}
+	v := r.Header.Hash()
+	r.hash.Store(v)
+	return v
+}
 
 // BatchRequest is used when requesting a range of batches from a peer.
 type BatchRequest struct {
-	Requester            string
-	EarliestMissingBatch *big.Int
+	Requester        string
+	CurrentHeadBatch *gethcommon.Hash // The requester's view of the current head batch, or nil if they haven't stored any batches.
```

Comment:
We moved from a model where the requester says which batches they're missing, to one where the sequencer works it out. The former had a fatal flaw, in that it failed when there had been a fork and it had rollups at the correct height, but on the wrong fork. Thanks Matt :)